SSA/Experimental/Bits/Fast/FiniteStateMachine.lean

```diff
@@ -679,3 +782,47 @@ theorem decideIfZeros_correct {arity : Type _} [DecidableEq arity]
     intro x s h
     use x
     exact h
+
+end FSM
+
+/--
+The fragment of predicate logic that we support in `bv_automata`.
+Currently, we support equality, conjunction, disjunction, and negation.
+This can be expanded to also support arithmetic constraints such as unsigned-less-than.
+-/
+inductive Predicate : Nat → Type _ where
+| eq (t1 t2 : Term) : Predicate ((max t1.arity t2.arity))
+| and  (p : Predicate n) (q : Predicate m) : Predicate (max n m)
+| or  (p : Predicate n) (q : Predicate m) : Predicate (max n m)
+-- For now, we can't prove `not`, because it needs NFA → DFA conversion
+-- the way Sid knows how to build it, or negation normal form,
+-- both of which is machinery we lack.
+-- | not (p : Predicate n) : Predicate n
+
+
+
+/--
+denote a reflected `predicate` into a `prop.
+-/
+def Predicate.denote : Predicate α → Prop
+| eq t1 t2 => t1.eval = t2.eval
+| and p q => p.denote ∧  q.denote
+| or p q => p.denote ∨  q.denote
```

Comment:
This is not quite right! We need to be taking in the arguments at the top-level, and pass those around.
Currently, we're denoting `eq t u` as equality of the *functions* that are denoted by `t` and `u`, meaning that we're universally quantifying each con-/disjunct separately.


```
def Predicate.denote (xs : α → BitStream) : Predicate α → Prop
| eq t1 t2 => t1.eval xs = t2.eval xs
| and p q => p.denote xs ∧ q.denote xs
| or p q => p.denote xs ∨ q.denote xs
```